crates/rollup/src/driver/mod.rs

```diff
@@ -118,12 +155,80 @@ where
     }
 
     /// Advance the pipeline to the next L2 block.
-    async fn step(&mut self, pipeline: &mut RollupPipeline<CP, BP, L2CP>) -> Result<()> {
-        let (l2_tip, l1_origin) = self.cursor.tip();
-        let _ = pipeline.step(l2_tip).await;
-        self.cursor.advance(l1_origin, l2_tip);
+    ///
+    /// Returns `true` if the pipeline can move forward again, `false` otherwise.
+    async fn step(&mut self, pipeline: &mut RollupPipeline<CP, BP, L2CP>) -> bool {
+        let l2_tip = self.cursor.tip();
+
+        match pipeline.step(l2_tip).await {
+            StepResult::PreparedAttributes => trace!("Perpared new attributes"),
+            StepResult::AdvancedOrigin => trace!("Advanced origin"),
+            StepResult::OriginAdvanceErr(err) => warn!("Could not advance origin: {:?}", err),
+            StepResult::StepFailed(err) => match err {
+                StageError::NotEnoughData => debug!("Not enough data to advance pipeline"),
+                _ => error!("Error stepping derivation pipeline: {:?}", err),
+            },
+        }
+
+        let derived_attributes = if let Some(attributes) = pipeline.peek() {
+            match self.validator.validate(attributes).await {
+                Ok(true) => {
+                    trace!("Validated payload attributes");
+                    pipeline.next().expect("Peeked attributes must be available")
+                }
+                Ok(false) => {
+                    error!("Failed payload attributes validation");
+                    // TODO: should we advance the pipeline iterator here?
```

Comment:
We'll probably want to allow the user to specify how they want to treat invalid payload attributes.

In the default case, I think it should probably log errors and continue.

Reply:
Awesome, will change the TODO to be a reminder about this future config option